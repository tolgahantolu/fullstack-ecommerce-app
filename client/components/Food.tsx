import Image from "next/image";
import Link from "next/link";
import React, { FormEvent, Key, useRef } from "react";
import useAddItemToCart from "../hooks/useAddItemToCart";
import FoodForm from "./FoodForm";

const Food: React.FC<{
  id: Object;
  title: String;
  desc: String;
  price: Number;
  kit: Boolean;
  category: String;
  ingredients: Array<string>;
  image: String;
  index: Key;
}> = ({ id, title, desc, price, kit, category, index, ingredients, image }) => {
  const amountInputRef = useRef<HTMLInputElement>(null);
  const { handlerAddItemToCart: handlerSubmit, amountIsValid } =
    useAddItemToCart(id, title, price, image, amountInputRef);

  return (
    <div
      key={index}
      className="col-span-1 w-full bg-theme-dark-grey rounded-[30px] py-3"
    >
      <div className="flex items-center justify-between gap-x-3 min-w-full min-h-[145px] max-h-[145px] h-full pl-4 pr-2 py-2">
        <div className="flex-1 flex flex-col h-full justify-between">
          <Link href={`/food/${id}`}>
            <h2
              data-testid="food-title"
              className="capitalize text-lg mb-1 font-medium"
            >
              {title}
            </h2>
            <div className="flex flex-row flex-wrap">
              {ingredients.length > 7 ? (
                <p className="text-theme-dark-grey2 text-sm font-medium leading-2">
                  {ingredients.slice(0, 7).join(", ").concat("...")}
                </p>
              ) : (
                <p className="text-theme-dark-grey2 text-sm font-medium leading-2">
                  {ingredients.join(", ")}
                </p>
              )}
            </div>
          </Link>
          <div className="mt-2 flex flex-row justify-between items-center">
            <p className="font-bold text-3xl leading-none">{`$${price}`}</p>
            <FoodForm
              onSubmit={(e: FormEvent) => handlerSubmit(e)}
              big={false}
              ref={amountInputRef}
            />
          </div>
          {!amountIsValid && (
            <p className="mt-4 text-left text-xs font-medium text-theme-dark-orange">
              Please enter a amount (1-5).
            </p>
          )}
        </div>
        <Image
          src={image as string}
          width={115}
          height={115}
          alt="product"
          className="object-cover object-center drop-shadow"
        />
      </div>
    </div>
  );
};

export default Food;
