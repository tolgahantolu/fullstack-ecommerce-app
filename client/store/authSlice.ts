import { createSlice } from "@reduxjs/toolkit";

const authSlice = createSlice({
  name: "auth",
  initialState: {
    user: null,
  },
  reducers: {
    authUser: (state, action) => {
      state.user = action.payload.user;
    },
  },
});

export const { authUser } = authSlice.actions;
export default authSlice.reducer;
