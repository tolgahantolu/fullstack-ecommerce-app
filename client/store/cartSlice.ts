import { createSlice, current } from "@reduxjs/toolkit";

const initialState: any = {
  products: [],
  counter: 0,
  totalQuantity: 0,
  totalAmount: 0,
};

const cartSlice = createSlice({
  name: "cart",
  initialState,
  reducers: {
    addItemToCart(state, action) {
      state.counter = state.counter + action.payload.counter;

      const { id, title, price, image } = action.payload;

      const existingProduct = state.products.find(
        (product: { id: Object; title: String; price: Number }) =>
          product.id === id
      );
      console.log(current(state.products));

      if (!existingProduct) {
        state.products.push({
          id: id,
          title: title,
          price: price,
          image: image,
          totalPrice: price * action.payload.amount,
          quantity: 0 + action.payload.amount,
        });
      } else {
        if (action.payload.amount) {
          existingProduct.quantity =
            existingProduct.quantity + action.payload.amount;
          existingProduct.totalPrice =
            existingProduct.totalPrice + price * action.payload.amount;
        } else {
          existingProduct.quantity++;
          existingProduct.totalPrice = existingProduct.totalPrice + price;
        }
      }

      // calc total amount
      state.totalAmount = state.products
        .map(
          (product: {
            id: Object;
            title: String;
            price: Number;
            totalPrice: Number;
            totalAmount: Number;
          }) => product.totalPrice
        )
        .reduce(
          (acc: any, totalPriceProduct: any) => acc + totalPriceProduct,
          0
        );

      console.log(current(state.products));
    },
    removeItemFromCart(state, action) {
      state.counter = state.counter - action.payload.counter;

      const { id, title, price } = action.payload;

      const existingProduct = state.products.find(
        (product: { id: Object; title: String; price: Number }) =>
          product.id === id
      );
      console.log(current(existingProduct));

      if (existingProduct.quantity === 1) {
        state.products = state.products.filter(
          (product: { id: Object }) => product.id != id
        );
      } else {
        existingProduct.quantity--;
        existingProduct.totalPrice = existingProduct.totalPrice - price;
      }

      // calc total amount
      state.totalAmount = state.products
        .map(
          (product: {
            id: Object;
            title: String;
            price: Number;
            totalPrice: Number;
            totalAmount: Number;
          }) => product.totalPrice
        )
        .reduce(
          (acc: any, totalPriceProduct: any) => acc + totalPriceProduct,
          0
        );
    },
  },
});

export const { addItemToCart, removeItemFromCart } = cartSlice.actions;
export default cartSlice.reducer;
