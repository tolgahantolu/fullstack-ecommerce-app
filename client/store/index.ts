import { configureStore } from "@reduxjs/toolkit";
import authSlice from "./authSlice";
import cartSlice from "./cartSlice";

const store = configureStore({
  reducer: {
    auth: authSlice,
    cart: cartSlice,
  },
});

export default store;
