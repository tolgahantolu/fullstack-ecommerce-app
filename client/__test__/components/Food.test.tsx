import React from "react";
import Food from "@/components/Food";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../../store";

const DUMMY_FOOD = {
  id: "63aa9a4b54282fbb206130a2",
  title: "veggie pizza kit",
  desc: "Lorem ipsum dolor sit amet.",
  price: 32,
  kit: false,
  category: "pizzas",
  ingredients: ["onion", "sausage"],
  image:
    "https://firebasestorage.googleapis.com/v0/b/nextjs-food-app-29e49.appspot.com/o/fpizza2.png?alt=media&token=1edfef55-6fd1-4cff-9621-85910e25663c",

  index: "0",
};

describe("Food Component", () => {
  beforeAll(() => {
    render(
      <Provider store={store}>
        <Food {...DUMMY_FOOD} />
      </Provider>
    );
  });

  test("render food data title and price correctly", () => {
    expect(screen.getByTestId("food-title")).toHaveTextContent(
      "veggie pizza kit"
    );
    expect(DUMMY_FOOD.price).toBeGreaterThan(0);
  });

  test("render food image correctly", () => {
    expect(DUMMY_FOOD.image).toMatchSnapshot();
  });
});
