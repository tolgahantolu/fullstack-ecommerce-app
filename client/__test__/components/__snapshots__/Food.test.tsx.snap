// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Food Component render food image correctly 1`] = `"https://firebasestorage.googleapis.com/v0/b/nextjs-food-app-29e49.appspot.com/o/fpizza2.png?alt=media&token=1edfef55-6fd1-4cff-9621-85910e25663c"`;
